### ISSUE TITLE
Shipping extensions with GLES names can break existing content

### ISSUE THREAD

**Issue by kainino0x** [KhronosGroup/WebGL#3604]
Please see this bug for details: https://github.com/emscripten-core/emscripten/issues/20798

tl;dr:
- Emscripten exposes the GL_EXTENSIONS directly from WebGL.
- Emscripten does not implement glClipControlEXT or glPolygonOffsetClampEXT
- Browser ships EXT_clip_control or EXT_polygon_offset_clamp
- Existing pages on the Web that were compiled using Emscripten, from C/C++ code that supported these extensions ([like Filament](https://github.com/google/filament/blob/e7c67d1adb45182a0b7e765a7d90564c26640da9/filament/backend/src/opengl/OpenGLContext.cpp#L281)) starts crashing

I have a few ideas about this but my suggestion on that bug is for the WebGL group to simply not ship new `EXT_` extensions that add new functions, and instead rebrand them as `WEBGL_`.

**Comment by kdashg** [comment]
Emscripten should be allowlisting extensions, and only adding them to its extension string if they are recognized and implemented. It is a defect in emscripten that it hacks together an extension string like this.

**Comment by kainino0x** [comment]
> Emscripten should be allowlisting extensions, and only adding them to its extension string if they are recognized and implemented. It is a defect in emscripten that it hacks together an extension string like this.

I agree, but realistically there is a bunch of content out there that has this bug and is not going to get rebuilt with a new version of Emscripten. And unfortunately I don't think we could really do browser telemetry to determine what sites depend on this behavior in order to do a proper "deprecate and remove" process.

> Are apps not being careful here?

Filament, at least, is not, and that is a pretty well battle-tested engine (on Android mostly), so I guess they have not had issues with it. AFAIK the presence of the string is supposed to guarantee the presence of the function(s), so this would only be needed if there's a driver bug?

**Comment by greggman** [comment]
If I understand the issue

```C++
  // Original native C/C++ code. Runs on native OpenGL ES

  bool supportsEXTClipControl = !!strstr(glGetString(GL_EXTENSIONS), "GL_EXT_clip_control");
  
  ...

  if (supportsEXTClipControl) {
    glClipControlEXT(...);             // crashes on emscripten since glClipControlEXT 
                                       // is not connected to anything. Or, if it's 
                                       // connected to a no-op then it's not actually 
                                       // doing the thing advertised as supported.
  }
```

**Comment by kainino0x** [comment]
Correct, I have some sample code in the Emscripten bug too:
https://github.com/kainino0x/emscripten-extension-bug/blob/main/main.cpp

**Comment by kdashg** [comment]
Driver bugs are a reality, and we've hit this class of 'em before. That's why Firefox is paranoid about it.
I do not want to change prefixes because of a poor reason like this, so I'd like to consider other avenues before deciding.

**Before moving forward with any changes to our process, I must insist that Emscripten fix its defect here before it causes further harm.**

**Comment by MarkCallow** [comment]
That PR does not fix the problem. It disables GetProcAddress thus preventing people from encountering the problem but in the process breaks any app that really needs to use GetProcAddress. A new compile option has to be specified to have GetProcAddress. There is a new PR #20948 <https://github.com/emscripten-core/emscripten/pull/20948> to flip the flag so GetProcAddress is available by default and the flag disables it. I do not understand why they are reluctant to fix the extension list they advertise. There is mention of some people wanting to see the unfiltered WebGL list but I don’t truly understand why. Possibly something to do with extensions that have no functions. But a filtered list could still include those extensions.

Regards

    -Mark
